File: web/features/coding/shared/sessionManager/detail/domain/messageBlocks.ts
import type { SessionMessage, SessionMessageBlock } from '../../types';

import { hasSessionCommandTags, parseSessionCommandTags } from './commandTags';
import { getToolDisplayName } from './toolCatalog';

export function getMessageBlocks(message: SessionMessage): SessionMessageBlock[] {
  if (message.blocks && message.blocks.length > 0) {
    return message.blocks;
  }

  if (!message.content.trim()) {
    return [];
  }

  return [{
    kind: message.role.toLowerCase() === 'system' ? 'system' : 'text',
    text: message.content,
    variant: message.role.toLowerCase() === 'system' ? 'system' : undefined,
  }];
}

export function isToolBlock(block: SessionMessageBlock): boolean {
  return block.kind === 'tool_call'
    || block.kind === 'tool_result'
    || block.kind === 'tool_execution';
}

export function getBlockText(block: SessionMessageBlock): string {
  const parts = [
    block.title,
    block.text,
    block.toolName,
    block.normalizedToolName,
    valueToSearchText(block.input),
    valueToSearchText(block.output),
    valueToSearchText(block.metadata),
  ];

  return parts.filter(Boolean).join('\n');
}

export function getMessagePreview(message: SessionMessage, maxLength = 96): string {
  const blocks = getMessageBlocks(message);
  const text = blocks
    .map((block) => {
      if (isToolBlock(block)) {
        return getToolPreview(block);
      }
      if (hasSessionCommandTags(block.text || '')) {
        const parsedCommand = parseSessionCommandTags(block.text || '');
        return [
          parsedCommand.commandName,
          parsedCommand.commandMessage,
          parsedCommand.commandArgs,
          parsedCommand.remainingText,
        ].filter(Boolean).join(' ');
      }
      return block.text || '';
    })
    .filter(Boolean)
    .join(' ');

  const collapsed = collapsePreview(text || message.content);
  if (collapsed.length <= maxLength) {
    return collapsed;
  }
  return `${collapsed.slice(0, maxLength)}...`;
}

export function isMeaningfulPreview(value: string | undefined): boolean {
  const normalized = collapsePreview(value || '').toLowerCase();
  if (!normalized) {
    return false;
  }
  if (normalized === 'unknown' || normalized === 'tool: unknown') {
    return false;
  }
  if (/^\[?tool:\s*unknown\]?$/i.test(normalized)) {
    return false;
  }
  if (/^\((user|assistant|tool|system) message\)$/i.test(normalized)) {
    return false;
  }
  return true;
}

export function valueToSearchText(value: unknown): string {
  if (value === null || value === undefined) {
    return '';
  }
  if (typeof value === 'string') {
    return value;
  }
  if (typeof value === 'number' || typeof value === 'boolean') {
    return String(value);
  }
  if (Array.isArray(value)) {
    return value.map(valueToSearchText).filter(Boolean).join('\n');
  }
  if (typeof value === 'object') {
    const record = value as Record<string, unknown>;
    const preferred = ['text', 'content', 'stdout', 'stderr', 'output', 'error', 'file_path', 'path', 'command', 'query', 'pattern'];
    const direct = preferred
      .map((key) => valueToSearchText(record[key]))
      .filter(Boolean)
      .join('\n');

    if (direct) {
      return direct;
    }

    try {
      return JSON.stringify(value, null, 2);
    } catch {
      return String(value);
    }
  }

  return '';
}

function getToolPreview(block: SessionMessageBlock): string {
  const directName = meaningfulName(block.toolName);
  if (directName) {
    return directName;
  }

  const normalizedName = meaningfulName(block.normalizedToolName);
  if (normalizedName) {
    return getToolDisplayName(normalizedName, undefined);
  }

  return meaningfulName(block.title) || '';
}

function meaningfulName(value: string | undefined): string {
  const collapsed = collapsePreview(value || '');
  return isMeaningfulPreview(collapsed) ? collapsed : '';
}

function collapsePreview(value: string): string {
  return value
    .replace(/<[^>]*>/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

File: tauri/src/coding/session_manager/message_blocks.rs
use std::collections::HashMap;

use serde_json::Value;

use super::tool_normalizer::{infer_tool_status, infer_tool_variant, normalize_tool_name};
use super::{SessionMessage, SessionMessageBlock};

pub(super) fn text_block(text: impl Into<String>) -> SessionMessageBlock {
    block_with_text("text", text)
}

pub(super) fn thinking_block(text: impl Into<String>) -> SessionMessageBlock {
    let mut block = block_with_text("thinking", text);
    block.variant = Some("thinking".to_string());
    block.title = Some("Thinking".to_string());
    block
}

pub(super) fn tool_call_block(
    tool_id: Option<String>,
    tool_name: impl Into<String>,
    input: Option<Value>,
) -> SessionMessageBlock {
    let tool_name = tool_name.into();
    let normalized_tool_name = normalize_tool_name(&tool_name);
    let mut block = empty_block("tool_call");
    block.variant = Some(infer_tool_variant(&normalized_tool_name, &tool_name));
    block.tool_id = tool_id;
    block.tool_name = Some(tool_name);
    block.normalized_tool_name = Some(normalized_tool_name);
    block.status = Some("pending".to_string());
    block.input = input;
    block
}

pub(super) fn tool_result_block(
    tool_id: Option<String>,
    tool_name: Option<String>,
    output: Option<Value>,
    is_error: Option<bool>,
) -> SessionMessageBlock {
    let normalized_tool_name = tool_name
        .as_deref()
        .map(normalize_tool_name)
        .unwrap_or_else(|| "unknown".to_string());
    let mut block = empty_block("tool_result");
    block.variant = Some(infer_tool_variant(
        &normalized_tool_name,
        tool_name.as_deref().unwrap_or(""),
    ));
    block.tool_id = tool_id;
    block.tool_name = tool_name;
    block.normalized_tool_name = Some(normalized_tool_name);
    block.output = output;
    block.is_error = is_error;
    block.status = Some(infer_tool_status(&block));
    block
}

pub(super) fn message_from_blocks(
    role: impl Into<String>,
    ts: Option<i64>,
    blocks: Vec<SessionMessageBlock>,
) -> SessionMessage {
    let blocks = pair_tool_blocks(blocks);
    let content = flatten_blocks_for_content(&blocks);
    SessionMessage {
        role: role.into(),
        content,
        ts,
        id: None,
        parent_id: None,
        message_type: None,
        blocks,
        model: None,
        usage: None,
        duration_ms: None,
        cost_usd: None,
        is_sidechain: None,
        metadata: None,
    }
}

pub(super) fn flatten_blocks_for_content(blocks: &[SessionMessageBlock]) -> String {
    blocks
        .iter()
        .filter_map(block_to_text)
        .filter(|text| !text.trim().is_empty())
        .collect::<Vec<_>>()
        .join("\n")
}

pub(super) fn pair_tool_blocks(blocks: Vec<SessionMessageBlock>) -> Vec<SessionMessageBlock> {
    let mut result = Vec::new();
    let mut pending_tool_call_indices: HashMap<String, usize> = HashMap::new();

    for block in blocks {
        if block.kind == "tool_call" {
            if let Some(tool_id) = block.tool_id.clone() {
                pending_tool_call_indices.insert(tool_id, result.len());
            }
            result.push(block);
            continue;
        }

        if block.kind == "tool_result" {
            if let Some(tool_id) = block.tool_id.as_deref() {
                if let Some(call_index) = pending_tool_call_indices.remove(tool_id) {
                    if let Some(call_block) = result.get_mut(call_index) {
                        call_block.kind = "tool_execution".to_string();
                        call_block.output = block.output.clone();
                        call_block.is_error = block.is_error;
                        call_block.status = block.status.clone();
                        call_block.status = Some(infer_tool_status(call_block));
                        if call_block.tool_name.is_none() {
                            call_block.tool_name = block.tool_name.clone();
                        }
                        if call_block.normalized_tool_name.is_none() {
                            call_block.normalized_tool_name = block.normalized_tool_name.clone();
                        }
                        continue;
                    }
                }
            }
        }

        result.push(block);
    }

    result
}

fn block_with_text(kind: &str, text: impl Into<String>) -> SessionMessageBlock {
    let mut block = empty_block(kind);
    block.text = Some(text.into());
    block
}

fn empty_block(kind: &str) -> SessionMessageBlock {
    SessionMessageBlock {
        kind: kind.to_string(),
        text: None,
        title: None,
        variant: None,
        language: None,
        tool_id: None,
        tool_name: None,
        normalized_tool_name: None,
        status: None,
        is_error: None,
        input: None,
        output: None,
        metadata: None,
    }
}

fn block_to_text(block: &SessionMessageBlock) -> Option<String> {
    match block.kind.as_str() {
        "tool_call" | "tool_execution" => {
            let name = block
                .tool_name
                .as_deref()
                .or(block.normalized_tool_name.as_deref())
                .unwrap_or("unknown");
            Some(format!("[Tool: {name}]"))
        }
        "tool_result" => block.output.as_ref().and_then(value_to_text),
        _ => block
            .text
            .clone()
            .or_else(|| block.output.as_ref().and_then(value_to_text)),
    }
}

fn value_to_text(value: &Value) -> Option<String> {
    match value {
        Value::Null => None,
        Value::String(text) => Some(text.to_string()),
        Value::Array(items) => {
            let text = items
                .iter()
                .filter_map(value_to_text)
                .filter(|text| !text.trim().is_empty())
                .collect::<Vec<_>>()
                .join("\n");
            if text.trim().is_empty() {
                Some(serde_json::to_string_pretty(value).unwrap_or_else(|_| value.to_string()))
            } else {
                Some(text)
            }
        }
        Value::Object(map) => {
            if let Some(text) = map.get("text").and_then(Value::as_str) {
                return Some(text.to_string());
            }
            if let Some(content) = map.get("content") {
                if let Some(text) = value_to_text(content) {
                    return Some(text);
                }
            }
            Some(serde_json::to_string_pretty(value).unwrap_or_else(|_| value.to_string()))
        }
        _ => Some(value.to_string()),
    }
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::*;

    #[test]
    fn pair_tool_blocks_merges_matching_call_and_result() {
        let blocks = pair_tool_blocks(vec![
            tool_call_block(
                Some("tool-1".to_string()),
                "Bash",
                Some(json!({ "command": "echo hi" })),
            ),
            tool_result_block(
                Some("tool-1".to_string()),
                Some("Bash".to_string()),
                Some(json!({ "stdout": "hi" })),
                None,
            ),
        ]);

        assert_eq!(blocks.len(), 1);
        assert_eq!(blocks[0].kind, "tool_execution");
        assert_eq!(blocks[0].normalized_tool_name.as_deref(), Some("bash"));
        assert_eq!(blocks[0].status.as_deref(), Some("success"));
    }

    #[test]
    fn pair_tool_blocks_does_not_merge_without_id() {
        let blocks = pair_tool_blocks(vec![
            tool_call_block(None, "Read", Some(json!({ "file_path": "Cargo.toml" }))),
            tool_result_block(None, Some("Read".to_string()), Some(json!("ok")), None),
        ]);

        assert_eq!(blocks.len(), 2);
        assert_eq!(blocks[0].kind, "tool_call");
        assert_eq!(blocks[1].kind, "tool_result");
    }

    #[test]
    fn flatten_blocks_keeps_text_and_tool_preview() {
        let message = message_from_blocks(
            "assistant",
            None,
            vec![
                text_block("hello"),
                tool_call_block(
                    Some("tool-1".to_string()),
                    "Grep",
                    Some(json!({ "pattern": "x" })),
                ),
            ],
        );

        assert!(message.content.contains("hello"));
        assert!(message.content.contains("[Tool: Grep]"));
    }
}
